nextcord/http.py

```diff
@@ -240,151 +707,280 @@ async def request(
+        # If a global rate limit for this authorization doesn't exist yet, make it.
+        if (global_rate_limit := self._global_rate_limits.get(auth)) is None:
+            global_rate_limit = self._make_global_rate_limit(auth, self._max_global_requests)
 
-        if not self._global_over.is_set():
-            # wait until the global lock is complete
-            await self._global_over.wait()
+        # If a rate limit for this url path doesn't exist yet, make it.
+        if (url_rate_limit := self._get_url_rate_limit(route.method, route, auth)) is None:
+            url_rate_limit = self._make_url_rate_limit(route.method, route, auth)
 
+        max_retry_count = 5
+        rate_limit_path = (
+            route.method,
+            route.bucket,
+            _get_logging_auth(auth),
+        )  # Only use this for logging.
+        ret: Union[Optional[str], dict] = None
         response: Optional[aiohttp.ClientResponse] = None
-        data: Optional[Union[Dict[str, Any], str]] = None
-        await lock.acquire()
-        with MaybeUnlock(lock) as maybe_lock:
-            for tries in range(5):
-                if files:
-                    for f in files:
-                        f.reset(seek=tries)
-
-                if form:
-                    form_data = aiohttp.FormData(quote_fields=False)
-                    for params in form:
-                        form_data.add_field(**params)
-                    kwargs["data"] = form_data
-
-                try:
-                    async with self.__session.request(method, url, **kwargs) as response:
+
+        # If retry_request is False and any of the rate limits are locked, don't continue and raise immediately.
+        if retry_request is False:
+            if global_rate_limit.locked:
+                _log.info(
+                    "Path %s was called with retry_request=False while the global rate limit is locked.",
+                    rate_limit_path,
+                )
+                raise HTTPInternalRatelimitLocked(
+                    "Request would exceed the global ratelimit and retries are disabled."
+                )
+
+            if url_rate_limit.locked:
+                _log.info(
+                    "Path %s was called with retry_request=False while the URL rate limit is locked.",
+                    rate_limit_path,
+                )
+                raise HTTPInternalRatelimitLocked(
+                    "Request would exceed the route ratelimit and retries are disabled."
+                )
+
+        # The loop is to allow migration to a different RateLimit if needed.
+        # If we hit this loop max_retry_count times, something is wrong. Either we're migrating buckets way
+        #  too much, 429s keep getting hit, or something is internally wrong.
+        for retry_count in range(max_retry_count):  # To prevent infinite loops.
+            should_retry = False
+            try:
+                async with global_rate_limit, url_rate_limit:
```

Comment:
This causes slowdowns.
Quick fix: swap them around
Long fix: wait for both to be ready before acquiring. Will happily send you a example of how to do this from my rate limit library.